eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs

```diff
@@ -243,12 +254,49 @@ actionWellFormed ga = failureUnless isWellFormed $ MalformedProposal ga
       ParameterChange _ ppd -> ppuWellFormed ppd
       _ -> True
 
+checkProposalsHaveAValidPrevious ::
+  forall era.
+  PrevGovActionIds era ->
+  GovSnapshots era ->
+  GovProcedures era ->
+  Test (ConwayGovPredFailure era)
+checkProposalsHaveAValidPrevious prevGovActionIds snapshots procedures =
+  let isValidMember ::
+        StrictMaybe (PrevGovActionId p (EraCrypto era)) ->
+        (GovAction era -> Bool) ->
+        (PrevGovActionIds era -> StrictMaybe (PrevGovActionId p (EraCrypto era))) ->
+        Bool
+      isValidMember prev snapshotCond prevActionAccessor =
+        all
+          (== Just True) -- lookup has to succeed _and_ purpose of lookedup action has to match condition
+          ( fmap (snapshotCond . gasAction)
+              . (flip Map.lookup . snapshotGovActionStates $ snapshots ^. curGovSnapshotsL @era)
+              . unPrevGovActionId
+              <$> prev
+          )
```

Comment:
This looks correct, but it is written in a very obscure way. I suggest rewriting it with regular pattern matching and clarifying the logic a bit